### PR TITLE
[CALCITE-3782]Bitwise agg operator Bit_And, Bit_OR and Bit_XOR support binary and varbinary type

### PR DESCRIPTION
**Pull request by wangxlong**
According to the discussion [CALCITE-3732](https://issues.apache.org/jira/browse/CALCITE-3732) , We should make bitwise operators work on all integer types, BINARY and VARBINARY. So Bit_And, Bit_OR and Bit_XOR agg operator should also support BINARY and VARBINARY.

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/runtime/CalciteResource.java [hunk 1]
```diff
@@ -918,4 +918,7 @@
 
   @BaseMessage("Invalid input for EXTRACTVALUE: xml: ''{0}'', xpath expression: ''{1}''")
   ExInst<CalciteException> invalidInputForExtractValue(String xml, String xpath);
+
+  @BaseMessage("Invalid length for bitwise operator of input binary value")
+  ExInst<CalciteException> invalidInputForBitwiseBinaryValue();
```
Comment (danny0405) [inline core/src/main/java/org/apache/calcite/runtime/CalciteResource.java]: Actually the error throws because the two params have different length.

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java [hunk 2]
```diff
@@ -1074,23 +1075,83 @@ private static RuntimeException notComparable(String op, Object b0,
...
     return b0 & b1;
   }
 
+  /** Helper function for implementing <code>BIT_AND</code> applied to binary values */
+  public static ByteString bitAnd(ByteString b0, ByteString b1) {
+
+    if (b0.length() == 0 || b1.length() == 0) {
+      return b1.length() == 0 ? b0 : b1;
+    }
+
```
Comment (danny0405) [inline core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java]: The code below seems more straight-forward.
```java
if (b0.length() == 0) {
      return b1;
}

if (b1.length() == 0) {
    return b0;
}
```

File: core/src/test/java/org/apache/calcite/sql/test/SqlOperatorBaseTest.java [hunk 3]
```diff
@@ -9074,6 +9083,11 @@ protected void checkAggType(SqlTester tester, String expr, String type) {
         false);
     final String[] values = {"1", "2", "2"};
     tester.checkAgg("bit_or(x)", values, 3, 0);
+    final String[] binaryValues = {
+        "CAST(x'01' AS BINARY)",
+        "cast(x'02' as BINARY)",
+        "cast(x'02' AS BINARY)"};
```
Comment (danny0405) [inline core/src/test/java/org/apache/calcite/sql/test/SqlOperatorBaseTest.java]: Add a test for `null` value.

Comment (danny0405) [reply core/src/test/java/org/apache/calcite/sql/test/SqlOperatorBaseTest.java]: We may also need to update the `reference.md` for the function signatures change.

**Comment by wangxlong** [comment]
@danny0405 Thanks for your advice, I have updated.

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java [hunk 4]
```diff
@@ -1074,23 +1075,92 @@ private static RuntimeException notComparable(String op, Object b0,
...
+    if (b0.length() != b1.length()) {
+      throw RESOURCE.differentLengthForBitwiseOperands(b0.length(), b1.length()).ex();
+    }
+    byte[] bytes0 = b0.getBytes();
+    byte[] bytes1 = b1.getBytes();
+    byte[] result = new byte[b0.length()];
+
+    IntStream.range(0, bytes0.length).forEach(i -> {
```
Comment (ritesh-kapoor) [inline core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java]: Should we move it to utility function. Seems like repetitive.

Comment (wangxlong) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:1129]: Sorry for reply later. The variable definition of the function is the same, but the main processing mode is different between BIT_AND, BIT_OR, BIT_XOR which are &, | and ^. So We don't need to move it to utility function.

**Comment by wangxlong** [comment]
@chunweilei Thanks for your advice. I have updated.

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java (hunk 4 above)
Comment (hsyuan) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:1129]: I think we should do.

Comment (wangxlong) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java]: Thanks @hsyuan @ritesh-kapoor. Yes, moved it to utility function looks more better. Thank you help me to have a look in your free time.

File: core/src/test/java/org/apache/calcite/sql/test/SqlOperatorBaseTest.java [hunk 5]
```diff
@@ -9096,6 +9114,12 @@ protected void checkAggType(SqlTester tester, String expr, String type) {
         false);
     final String[] values = {"1", "2", "1"};
     tester.checkAgg("bit_xor(x)", values, 2, 0);
+    final String[] binaryValues = {
+        "CAST(x'01' AS BINARY)",
+        "cast(x'02' as BINARY)",
+        "cast(x'01' AS BINARY)",
+        "cast(null AS BINARY)"};
+    tester.checkAgg("bit_xor(x)", binaryValues, "02", 0);
```
Comment (hsyuan) [inline core/src/test/java/org/apache/calcite/sql/test/SqlOperatorBaseTest.java:9122]: What if it has only 1 element?

Comment (hsyuan) [reply core/src/test/java/org/apache/calcite/sql/test/SqlOperatorBaseTest.java:9122]: Can you add the case?

Comment (wangxlong) [reply core/src/test/java/org/apache/calcite/sql/test/SqlOperatorBaseTest.java:9124]: Yes, I have added. When it has only 1 element, the result is itself.

Comment (hsyuan) [reply core/src/test/java/org/apache/calcite/sql/test/SqlOperatorBaseTest.java:9124]: Can you add a case with duplicate element and distinct for BIT_XOR? I am afraid it won't work.

Comment (wangxlong) [reply core/src/test/java/org/apache/calcite/sql/test/SqlOperatorBaseTest.java:9124]: You are right. Distinct for BIT_XOR isn't work. For SqlBitOpAggFunction#getDistinctOptionality is Optionality.IGNORED. So I think, for  BIT_AND and BIT_OR, We can set Optionality.IGNORED. And for BIT_XOR, We can set Optionality.OPTIONAL. What do you think of it?

Comment (hsyuan) [reply core/src/test/java/org/apache/calcite/sql/test/SqlOperatorBaseTest.java:9124]: Correct.

File: core/src/main/java/org/apache/calcite/sql/fun/SqlBitOpAggFunction.java [hunk 6]
```diff
@@ -62,6 +62,17 @@ public SqlBitOpAggFunction(SqlKind kind) {
...
   @Override public Optionality getDistinctOptionality() {
-    return Optionality.IGNORED;
+    final Optionality optionality;
+
+    switch (kind) {
+    case BIT_AND:
+    case BIT_OR:
+      optionality = Optionality.IGNORED;
+      break;
+    default:
```
Comment (hsyuan) [inline core/src/main/java/org/apache/calcite/sql/fun/SqlBitOpAggFunction.java:72]: Can you add BIT_XOR before default?

Comment (wangxlong) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlBitOpAggFunction.java:72]: Yes, It is ok.

File: core/src/main/java/org/apache/calcite/runtime/BitwiseFunctions.java [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+
+  /**
+   * Util for bitwise function applied to two byteString values.
+   * @param b0 The first byteString value operand of bitwise function.
+   * @param b1 The second byteString value operand of bitwise function.
+   * @param operator BitWise Operator.
+   * @return
+   */
+  private static ByteString binaryOperator(
+      ByteString b0, ByteString b1, BitwiseOperators operator) {
```
Comment (hsyuan) [inline core/src/main/java/org/apache/calcite/runtime/BitwiseFunctions.java:75]: Can we use `Supplier` like https://github.com/apache/calcite/blob/master/core/src/main/java/org/apache/calcite/plan/RelTraitSet.java#L232 instead of BitwiseOperators enum?

Comment (wangxlong) [reply core/src/main/java/org/apache/calcite/runtime/BitwiseFunctions.java]: Of course. Thank you for your suggestion. Updates.

Comment (hsyuan) [reply core/src/main/java/org/apache/calcite/runtime/BitwiseFunctions.java]: Can you cherry-pick this patch https://github.com/hsyuan/calcite/commit/af0d164eeaa615e489c58439afc3d109d98e7810 to your pull request?

Comment (wangxlong) [reply core/src/main/java/org/apache/calcite/runtime/BitwiseFunctions.java]: Yes, updated. Thank you @hsyuan help very much.

File: core/src/main/resources/org/apache/calcite/runtime/CalciteResource.properties [hunk 8]
```diff
@@ -301,4 +301,5 @@ InvalidInputForXmlTransform=Invalid input for XMLTRANSFORM xml: ''{0}''
 InvalidInputForExtractValue=Invalid input for EXTRACTVALUE: xml: ''{0}'', xpath expression: ''{1}''
 InvalidInputForExtractXml=Invalid input for EXTRACT xpath: ''{0}'', namespace: ''{1}''
 InvalidInputForExistsNode=Invalid input for EXISTSNODE xpath: ''{0}'', namespace: ''{1}''
+DifferentLengthForBitwiseOperands=Different length for bitwise operands: the first: {0,number,#}, the second: {1,number,#}
```
Comment (hsyuan) [inline core/src/main/resources/org/apache/calcite/runtime/CalciteResource.properties:304]: @wangxlong I am planning to get this into next release.
Can you add a negative test case to show this error message?
Thanks

Comment (hsyuan) [reply core/src/main/resources/org/apache/calcite/runtime/CalciteResource.properties:304]: Also pelease rebase to the latest master.

Comment (wangxlong) [reply core/src/main/resources/org/apache/calcite/runtime/CalciteResource.properties:304]: @hsyuan Yes, Done.
